streambuffer.go

```diff
@@ -51,48 +72,69 @@ func (b *StreamBuffer) WriteAt(p []byte, off int64) (n int, err error) {
 	b.Lock()
 	defer b.Unlock()
 
+	// Check to see if the current write can fit into our current buffer window
+	// denoted by [b.offset, b.offset + buffer length * buffer size)
 	if off < b.offset {
 		return 0, ErrBeforeBounds
 	}
 	if off+int64(len(p)) > b.offset+int64(b.buffers.Len())*int64(b.bufferSize) {
 		return 0, ErrAfterBounds
 	}
+
+	// normalizedOffset is the offset of the write relative to the first buffer
+	// in our ring of buffers
 	normalizedOffset := off - b.offset
-	written := 0
+
+	// written is how much data we have written so far
+	var written int
 
 	buffers := b.buffers
 	for i := 0; i < b.buffers.Len(); i++ {
-		// if the write is beyond the current buffer then continue
+		// if the write is beyond the current buffer, then continue
 		if normalizedOffset+int64(written) > int64((i+1)*b.bufferSize) {
 			buffers = buffers.Next()
 			continue
 		}
+
 		// copy the portion of p bytes that fit in current buffer
 		buf := buffers.Value.(*bytes.Buffer)
 		data := buf.Bytes()
 
+		// bufferStart is the normalized offset of the buf above relative to
+		// the current offset of the overall buffer
 		bufferStart := i * b.bufferSize
-		bufferEnd := (i + 1) * b.bufferSize
 
-		copyStart := normalizedOffset + int64(written-bufferStart)
-		copyEnd := copyStart + int64(len(p)-written)
-		if copyEnd > int64(bufferEnd) {
-			copyEnd = int64(bufferEnd - bufferStart)
+		// copyStart and copyEnd are the destination boundaries of the copy
+		// operation into the current buffer like [copyStart, copyEnd)
+		copyStart := int(normalizedOffset + int64(written) - int64(bufferStart))
+		copyEnd := copyStart + len(p) - written
+
+		// Don't copy more than the size of the current buffer. If we're going
+		// to go beyond, limit it. We'll be sure to write to the next buffer
+		// next.
+		if copyEnd > b.bufferSize {
+			copyEnd = b.bufferSize
 		}
-		copy(data[copyStart:copyEnd], p[written:copyEnd-copyStart+int64(written)])
 
-		written += int(copyEnd - copyStart)
+		// Actually perform the copy operation
+		copyLength := copyEnd - copyStart
+		copy(data[copyStart:copyEnd], p[written:copyLength+written])
 
+		// If we've finished writing, go ahead and break out. If not, advance
+		// to the next write that we're to perform.
+		written += copyLength
 		if written >= len(p) {
 			break
 		}
-
 		buffers = buffers.Next()
 	}
 
-	// update state for written chunks
+	// Update the state for written chunks. We'll use this mapping later to
+	// figure out the largest contiguous chunk when we flush the data.
+	// FIXME: Subsequent writes to this offset of a different length will break
```

Comment:
yeah, this is true. this is another assumption that this package makes but that should generally be true for the use-case we are using it for so it may or may not actually need to be fixed.